### ISSUE TITLE
A live component's parent LiveView receives live_select's handle_info and handle_event messages

### ISSUE THREAD

**Issue by cblock** [maxmarcon/live_select#9]
First of all thanks for this very nice piece of software! It's filling a gap that I previously tried to work around using Select2 and some custom code. I highly appreciate the effort you put into this project.  When trying it out in a pet project of mine I found a potential bug:

To reproduce I've set up a minimal example app at: https://github.com/cblock/live_select_sample using Phoenix 1.7.0-rc.0. Then I added live_select 0.3.2 as per documentation.

For demonstration puposes let's assume that I want to use [live_select to choose a person's location](https://github.com/cblock/live_select_sample/blob/main/lib/live_select_sample_web/live/person_live/form_component.ex).

I implemented the `handle_info` callback [in the live_component itself](https://github.com/cblock/live_select_sample/blob/main/lib/live_select_sample_web/live/person_live/form_component.ex#L36) (not its parent liveviews which would be PersonLive.Index or PersonLive.Show respectievely depending on whether a user trigger's the form from the index or show view). 

When I start the server, I see the following log output: 

```
[debug] HANDLE EVENT
  Component: LiveSelect.Component
  View: LiveSelectSampleWeb.PersonLive.Index
  Event: "keyup"
  Parameters: %{"key" => "Control", "value" => "test"}
[debug] Replied in 146µs
[debug] warning: undefined handle_info in LiveSelectSampleWeb.PersonLive.Index. Unhandled message: %LiveSelect.ChangeMsg{id: "person_location_component", field: :location, text: "test", module: LiveSelect.Component}
[debug] HANDLE EVENT
  Component: LiveSelect.Component
  View: LiveSelectSampleWeb.PersonLive.Index
  Event: "blur"
  Parameters: %{"value" => "test"}
[debug] Replied in 49µs
```

To my understanding, live_select does not take into account the [form's `phx-target` attribute](https://github.com/cblock/live_select_sample/blob/main/lib/live_select_sample_web/live/person_live/form_component.ex#L20). Instead it always sends  messages to the live_component's parent live view.

IMHO this behavior is not desirable as it leads to code duplication: I'd have to implement callbacks both within Index and Show LiveView.

**Comment by maxmarcon** [comment]
Hi @cblock! Thank you very much for your appreciation and for opening this issue.

You're the second person I'm aware of who's trying to use live_select from within a LiveComponent. The problem is that messages (as opposed to events) can only be sent to LiveViews, and not to LiveComponents. LiveComponents don't have their own processes and as such don't implement the `handle_info` callback. The `phx-target` attribute can be used to route an event to a specific component, but there's no equivalent for messages.

So why is LiveSelect  using a message? That's because a LiveComponent can't programmatically send an event to its parent.
This is a big architectural flaw of LiveView IMO. If it were possible, LiveSelect could simply send an event to the parent (whatever this is, LiveComponent or LiveView) and we'd be done. Since this isn't possible I had to resort to messages.

I think your use case is important, as I feel that more and more people are using nested hierarchies of live components.

Maybe it's possible to get rid of `LiveSelect.ChangeMsg` altogether and turn the interaction with the parent into an event, triggered directly by the UI. So in the live_select text input you'd have:

```elixir
<%= text_input(@form, @text_input_field, 
phx_change: "live_select_change",
phx_target: @target,
phx_value_live_select_id: @id  # <-- don't know if this would work
...
%>
```
where `@target` is can optionally be passed as a parameter to `LiveSelect.live_select/3`.

(Alternatively, the LiveSelect hook could send the "live_select_change" event to the parent component using `pushEventTo`)

Then your parent LiveComponent or LiveView would have to handle the event:

```elixir
def handle_event("live_select_change", %{"value" => text, "live_select_id" => live_select_id}, socket) do 
    options = search_for_options(text)

    LiveSelect.update_options(id, options)

   {:noreply, socket}
end
``` 

('LiveSelect.update_options/2' would have to be modified or extended to accept an id)

Does this make sense? This are only my initial ideas, I haven't tested them yet. If you can think of alternatives, I'd love to hear about them.

**Comment by cblock** [comment]
Hi @maxmarcon and thanks for your quick reply.

I think it would be great if live_select seamlessly "plugs" into automatically generated code generated by Phoenix 1.7  `mix phx.gen.live` command.

I think your alternative idea points into that direction. So, if I understand that correctly we'd write something like this:


```elixir
defmodule LiveSelectSampleWeb.PersonLive.FormComponent do
  use LiveSelectSampleWeb, :live_component
  import LiveSelect

  alias LiveSelectSample.People

  @impl true
  def render(assigns) do
    ~H"""
    <div>
      <.header>
        <%= @title %>
        <:subtitle>Use this form to manage person records in your database.</:subtitle>
      </.header>

      <.simple_form
        :let={f}
        for={@changeset}
        id="person-form"
        phx-target={@myself}
        phx-change="validate"
        phx-submit="save"
      >
        <.input field={{f, :name}} type="text" label="name" />
        <%!-- remove automatically generated field and replace with <.live_select />
        <.input field={{f, :location}} type="text" label="location" /> 
        --%>
        <%= live_select(f, :location) %>
        <:actions>
          <.button phx-disable-with="Saving...">Save Person</.button>
        </:actions>
      </.simple_form>
    </div>
    """
  end

  @impl true
  def handle_event("live_select_change", %{"value" => text, id: "person_location_component"}, socket) do 
    options = search_for_options(text)
    live_select_change = %{id: "person_location_components", options: ["city name 1", "city name 2"] })
    {:noreply, assign(socket, :live_select_options, live_select_change}
  end

  @impl true
  def handle_event("validate", %{"person" => person_params}, socket) do
    changeset =
      socket.assigns.person
      |> People.change_person(person_params)
      |> Map.put(:action, :validate)

    {:noreply, assign(socket, :changeset, changeset)}
  end
end
```

At least in this specific scenario `phx-target` attribute on the live_select component would be unnecessary as `phx-target` is already specified on the form level (however there might be use cases outside forms where such an attribute might make sense).

Also I'd prefer to stick with an explicit way of coding the change (similar to how "validate" ist coded by default). So instead of an opaque  method call like `LiveSelect.update_options(id, options)` I'd prefer to something like

``` 
    live_select_change = %{id: "person_location_components", options: ["city name 1", "city name 2"] })
    {:noreply, assign(socket, :live_select_options, live_select_change}
```

Does that make sense?

**Comment by cblock** [comment]
...one more thought: With this approach we could also nicely handle the case, where a user initially focuses the live select and wo expect an unfiltered list of options so select from (which would be filtered down once he starts typing). We'd only have to add another 

``def handle_event("live_select_change", %{"value" => "", id: "person_location_component"}, socket) do```

that returns the initial (unfiltered) list of options.

**Comment by maxmarcon** [comment]
Thanks for the very interesting points. I think you convinced me that moving away from custom function like `update_options` is a good idea. Instead, as you say, the user could communicate with the component via assigns or the standard `send_update/3`, and live_select would only specify a convention on what assigns it will accept.

I like that. It would also fit well with one of the next planned changes (see TODO on the README), which is removing the `live_select/3` function and rendering live select using a Phoenix component:

```
<LiveSelect.live_select field={{form, field}} opts={opts} />
```

as this seems to be the direction LiveView forms are moving towards anyway. This is a topic however which is kind of independent from our original issue here.

It's worth noting that with the "assign-based" approach, we do not necessarily need the id anymore:

```
def handle_event("live_select_change", %{"value" => text}, socket) do
    options = list_all_options(text)

    {:noreply, assign(socket, :live_select_options, %{options: options}}
end
```

unless you want to use `send_update/3` in which case we could extend `live_select/3` to accept a custom id as an option, so that the user can set an id and use it later to pass it to `send_update/3`

All this fits also quite well with the other proposed enhancement https://github.com/maxmarcon/live_select/issues/8, which is about adding functionality to programmatically reset the selection. This again could be done via assigns.

All these are breaking changes from the current API so maybe suited for a 1.0.0 version

**Comment by goblinJoel** [comment]
I couldn't follow whether this event would go to the parent LiveComponent or to the grandparent LiveView in this case, but if it's going to the LiveView, it would still need to provide an id that the user can specify so that the LiveView can tell which of possibly many components it is. This is a use case I'm hitting at the moment: I have a list of items, each of which uses a LiveSelect but needs its own search criteria. So when I get a ChangeMsg, I need to know which component triggered it so I can search for things that component should be able to pick from.

**Comment by maxmarcon** [comment]
> I couldn't follow whether this event would go to the parent LiveComponent or to the grandparent LiveView

If the form has a `phx-target` attribute set, it would go to the subject of `phx-target`. If there's no phx-target, it would go to the LiveView.
The rationale for using events instead of messages is to allow LiveSelect to talk to the parent even if it's a LiveComponent.

> it would still need to provide an id that the user can specify so that the LiveView can tell which of possibly many components it is.

You could use form and field names (available under `_target`) to discriminate between live select components:

```elixir
def handle_event("live_select_change", %{"_target" => [form_name,  :assoc_id]} = params, socket) do
  options = search_for_form(form_name, params[form_name]["assoc_id_text_input"])
 
  # since we're passing the options to LiveSelect via assigns and not with update_options anymore, 
  # all you have to do is update the right assigns in your LiveView or LiveComponent
end
```

I think that sending the LiveSelect id as part of the change event would be nice, but it isn't trivial. I tried something like this:

```
<%= text_input(@form, @text_input_field, 
phx_change: "live_select_change",
phx_value_live_select_id: @id
...
%>
```

and `live_select_id` is not being sent as part of the event params. Looks like the `phx-change` event doesn't support additional custom values. Happy to be proven wrong!

Is there any other way to pass the id?

**Comment by KristerV** [comment]
third person here. i'm sure with the release of phoenix 1.7 this will become common (i'm on 1.7) because of how phx.gen structures the app (seen above). right now i'm going to have to duplicate the handle_info in two liveviews (index & edit). not a huge issue i might add. just a nitpick to a great library (thanks!).

**Comment by maxmarcon** [comment]
thanks @KristerV , much appreciated!

Yours is not a nitpick but a very good point. I'm planning to ditch the old update cycle based on `LiveSelect.ChangeMsg` and `LiveSelect.update_options/2` in favor of using simple events and `LiveView.send_update/3`.

This would allow putting the update logic in a LiveComponent, like the form components generated by phoenix 1.7

I think I figured out how to do it and will start working on it soon. So the usage will become something like:

```elixir
<.live_select form={form} field={field} id={custom_id} />
```

and the update handle in the LV or LC will be:

```elixir
def handle_event("live_select_update", %{"id" => component_id, "text" => text}, socket) do 
   options = get_new_options(text)

   send_update(LiveSelect.Component, id: component_id, options: options)

   {:noreply, socket}
end
```

As I mentioned above, this will be a breaking change for the 1.0.0 release.

**Comment by KristerV** [comment]
Breaking change is totally okay and expected.

I opted for the default select behaviour, because i couldn't get live_select working properly. Mainly the dropdown never hid itself and clicking on the select made current content disappear.

It didn't make sense to report errors if there's a major rework coming anyway. Will try again with 1.7 support.

**Comment by maxmarcon** [comment]
> I opted for the default select behaviour, because i couldn't get live_select working properly. Mainly the dropdown never hid itself and clicking on the select made current content disappear.

could you clarify what you mean by "default select behaviour" ? Are the issues you're mentioning preventing you from using the library? If so, I'd like you to report it. There's gonna me some rework coming up, but the library is supposed to be working already.

**Comment by KristerV** [comment]
> could you clarify what you mean by "default select behaviour" ?
yeah i ditched this lib and went with just the HTML version for now.

since i already took out the library from my project i went ahead and created a new one just to demo the situation. and well, the bug isn't showing up anymore.. as far as i know i did everything in exactly the same way.. so must be a conflict in my own project. i will try this again, but will wait for new release. i'll report my findings then.

thanks for the quick replies btw!

**Comment by maxmarcon** [comment]
I'm glad to announce that his issue is finally resolved now that version 1.0.0 is released 🎉 

Now it's really easy to use LiveSelect form a live component. First of all, you render LiveSelect as a function component:

```elixir
<.live_select form={form} field={field} mode={:tags} /
```

Then, you handle input changes in your LiveComponent (or LiveView) with `handle_event/3`:

```
def handle_event("live_select_change", %{"text" => text, "id" => live_select_id}, socket) do
    options = retrieve_options(text)
    
    send_update(LiveSelect.Component, id: live_select_id, options: options)
    
    {:noreply, socket}
end
```

It would be great if some of the folks that were using LiveSelect in a LiveComponent could try this out.

**Comment by maxmarcon** [comment]
closed with 295531025afa144912daa53b12670c0fcf830bc9
